pkg/log/zap_async_buffered_write_core.go

```diff
@@ -0,0 +1,225 @@
+		FlushInterval: cfg.AsyncWriteFlushInterval,
+	}
+	asyncTextIOCore := &asyncTextIOCore{
+		LevelEnabler:        enab,
+		notifier:            syncutil.NewAsyncTaskNotifier[struct{}](),
+		enc:                 enc,
+		bws:                 bws,
+		pending:             make(chan *entryItem, cfg.AsyncWritePendingLength),
+		writeDroppedTimeout: cfg.AsyncWriteDroppedTimeout,
+		stopTimeout:         cfg.AsyncWriteStopTimeout,
+		maxBytesPerLog:      cfg.AsyncWriteMaxBytesPerLog,
+	}
+	go asyncTextIOCore.background()
+	return asyncTextIOCore
+}
+
+// asyncTextIOCore is a wrapper around the textIOCore that writes the logs to the underlying buffered write syncer.
+type asyncTextIOCore struct {
+	zapcore.LevelEnabler
+
+	notifier            *syncutil.AsyncTaskNotifier[struct{}]
+	enc                 zapcore.Encoder
+	bws                 *zapcore.BufferedWriteSyncer
+	pending             chan *entryItem // the incoming new write requests
+	writeDroppedTimeout time.Duration
+	stopTimeout         time.Duration
+	maxBytesPerLog      int
+}
+
+// entryItem is the item to write to the underlying buffered write syncer.
+type entryItem struct {
+	buf   *buffer.Buffer
+	level zapcore.Level
+}
+
+// With returns a copy of the Core with the given fields added.
+func (s *asyncTextIOCore) With(fields []zapcore.Field) zapcore.Core {
+	enc := s.enc.Clone()
+	switch e := enc.(type) {
+	case *textEncoder:
+		e.addFields(fields)
+	case zapcore.ObjectEncoder:
+		for _, field := range fields {
+			field.AddTo(e)
+		}
+	default:
+		panic(fmt.Sprintf("unsupported encode type: %T for With operation", enc))
+	}
+	return &asyncTextIOCore{
+		LevelEnabler:        s.LevelEnabler,
+		notifier:            s.notifier,
+		enc:                 s.enc.Clone(),
+		bws:                 s.bws,
+		pending:             s.pending,
+		writeDroppedTimeout: s.writeDroppedTimeout,
+		stopTimeout:         s.stopTimeout,
+		maxBytesPerLog:      s.maxBytesPerLog,
+	}
+}
+
+// Check checks if the entry is enabled by the level enabler.
+func (s *asyncTextIOCore) Check(ent zapcore.Entry, ce *zapcore.CheckedEntry) *zapcore.CheckedEntry {
+	if s.Enabled(ent.Level) {
+		return ce.AddCore(ent, s)
+	}
+	return ce
+}
+
+// Write writes the underlying buffered write syncer and buffers the writes in a channel.
+// asyncTextIOCore doesn't promise the write operation is done.
+// the write operation will be dropped if the buffer is full or the underlying buffered write syncer is blocked.
+func (s *asyncTextIOCore) Write(ent zapcore.Entry, fields []zapcore.Field) error {
+	buf, err := s.enc.EncodeEntry(ent, fields)
+	if err != nil {
+		return err
+	}
+	entry := &entryItem{
+		buf:   buf,
+		level: ent.Level,
+	}
+
+	length := buf.Len()
+	if length == 0 {
+		return nil
+	}
+	select {
+	case s.pending <- entry:
+		metrics.LoggingPendingWriteLength.Inc()
+		metrics.LoggingPendingWriteBytes.Add(float64(length))
+	case <-time.After(s.writeDroppedTimeout):
+		metrics.LoggingDroppedWrites.Inc()
```

Comment:
This might lose log, shall we handle different level in differnt ways? Like sync for Fatal/Error, async for info/warning.

Reply:
Or at least fall back to stderr